completion/all.go

```diff
@@ -0,0 +1,94 @@
+package completion
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/coder/serpent"
+)
+
+const (
+	BashShell  string = "bash"
+	FishShell  string = "fish"
+	ZShell     string = "zsh"
+	Powershell string = "powershell"
+)
+
+var shellCompletionByName = map[string]func(io.Writer, string) error{
+	BashShell:  GenerateCompletion(bashCompletionTemplate),
+	FishShell:  GenerateCompletion(fishCompletionTemplate),
+	ZShell:     GenerateCompletion(zshCompletionTemplate),
+	Powershell: GenerateCompletion(pshCompletionTemplate),
+}
```

Comment:
nit: might be good to [lazy wrap]([url](https://github.com/coder/coder/blob/e4aef272fa1619d82abab1d011af1a5ab27011d5/coderd/util/lazy/value.go#L4)) these.

We've are often punished by libraries doing useless computation during init.

Reply:
a minor thing: it may not be worth caching the template output at all and always generating it on-demand. Less state means less code and less bugs. Library consumers can implement this caching on their own if they're generating tons of scripts.

Reply:
The computation on init here is already pretty minimal. We're not storing the template output as the map value, just a closure.